### PR TITLE
Make workflow the parent of its subtypes

### PR DESCRIPTION
**Pull request by netama**
In order to support upcoming changes in `deploy`, we need workflow to be the parent of its subtypes.
This includes:
* Changing the parent
* Adjusting the reference resolution to find the "grandparent" when needed

Tried to keep the changes minimal - not changing the nested paths of the workflow or its subtypes, and still supporting having fields with the same object types nested under "regular" types (so - only looking at the grandparent if the parent is the workflow).

@ori-moisis if there's a relevant ticket to link this to please add it.

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 1]
```diff
@@ -485,7 +484,8 @@ const fixDependentInstancesPathAndSetParent = (elements: Element[]): void => {
     if (customObject.path) {
       instance.path = [
         ...customObject.path.slice(0, -1),
-        ...(workflowDependentMetadataTypes.has(instance.elemID.typeName)
+        ...((workflowDependentMetadataTypes.has(instance.elemID.typeName)
+            || instance.elemID.typeName === WORKFLOW_METADATA_TYPE)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:488]: nit: seems like the issue exists in the code before this change, but this is comparing a Salto ID with a salesforce ID (should be comparing `metadataType` and not `elemID.typeName`)

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 2]
```diff
@@ -512,7 +512,10 @@ const fixDependentInstancesPathAndSetParent = (elements: Element[]): void => {
         return
       }
       setDependingInstancePath(instance, customObj)
-      addObjectParentReference(instance, customObj)
+      // workflow-dependent types already have the parent set
+      if (!workflowDependentMetadataTypes.has(metadataType(instance))) {
+        addObjectParentReference(instance, customObj)
+      }
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:518]: nit: again, not part of this change (but I do think this changes makes it a little worse because it makes it more complicated logic) but it seems like this should be handled in the workflow filter rather than have a bunch of workflow specific code in an unrelated filter

File: packages/salesforce-adapter/src/filters/field_references.ts [hunk 3]
```diff
@@ -102,15 +103,26 @@ const workflowActionMapper: ContextValueMapperFunc = (val: string) => {
...
   ReferenceContextStrategyName, ContextFunc
 > = {
   none: () => undefined,
   instanceParent: ({ instance, elemByElemID }) => {
-    const parent = makeArray(instance.annotations[INSTANCE_ANNOTATIONS.PARENT])[0]
-    return (isReferenceExpression(parent)
-      ? apiName(elemByElemID[parent.elemId.getFullName()])
-      : undefined)
+    let parent = getParentElement({ instance, elemByElemID })
+    if (isInstanceElement(parent) && metadataType(parent) === WORKFLOW_METADATA_TYPE) {
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/filters/field_references.ts:122]: I think this would be really confusing to read.
in the config it will read as "use the parent" but in fact it will not take the parent if the parent happens to be a workflow instance, and there is no reason for someone reading to config to suspect this.
seems like there are already workflow specific lookup functions here, maybe worth splitting this off to be a `parentOfParent` lookup?

**Review by ori-moisis — APPROVED** [review]

File: packages/salesforce-adapter/src/filters/custom_objects.ts [hunk 4]
```diff
@@ -512,7 +512,10 @@ const fixDependentInstancesPathAndSetParent = (elements: Element[]): void => {
         return
       }
       setDependingInstancePath(instance, customObj)
-      addObjectParentReference(instance, customObj)
+      // workflow-dependent types already have the parent set
+      if (!workflowDependentMetadataTypes.has(metadataType(instance))) {
```
Comment (omrilit) [inline packages/salesforce-adapter/src/filters/custom_objects.ts:516]: I think that we should stick with `instance.elemID.typeName` or `metadataType(instance)` but we tend to use both in this flow.

File: packages/salesforce-adapter/src/filters/field_references.ts [hunk 5]
```diff
@@ -102,15 +103,26 @@ const workflowActionMapper: ContextValueMapperFunc = (val: string) => {
   return workflowActionTypeMapping[val]
 }
 
+const getParentElement = ({ instance, elemByElemID }: {
+  instance: InstanceElement
+  elemByElemID: ElemIDToElemLookup
+}): Element | undefined => {
+  const parent = makeArray(instance.annotations[INSTANCE_ANNOTATIONS.PARENT])[0]
```
Comment (omrilit) [inline packages/salesforce-adapter/src/filters/field_references.ts:110]:
```suggestion
  const [parent] = makeArray(instance.annotations[INSTANCE_ANNOTATIONS.PARENT])
```

**Review by omrilit — APPROVED** [review]

**Comment by ori-moisis** [comment]
We should not merge this.
it will create a dependency cycle when trying to add a new workflow:
workflow has references to its nested instances, nested instances have parent reference back to the workflow, so it would be impossible to add a new workflow

not sure what we can do with this...

**Comment by ori-moisis** [comment]
closing to avoid accidental merge

File: packages/salesforce-adapter/src/filters/field_references.ts (hunk 5 above)
Comment (netama) [reply packages/salesforce-adapter/src/filters/field_references.ts:110]: what's the difference?
